docs/proposals/event-notification-api.md

```diff
@@ -0,0 +1,122 @@
+
+For the proposed API, an `Event` type is uniquely identified by a key computed as `{Event.InvolvedObject.Kind}:{Event.Reason}`.
+
+### API
+
+The Event Notification API provides:
+
+1. A way for notification providers to augment Origin with support for provider-defined notification types.
+2. A way for users to declare their notification preferences per event type.
+3. A way for Origin to augment Event types with useful presentation metadata.
+
+```go
+// EventNotificationType is a type of notification the cluster supports. Provider implementations
+// declare their support for notifications by creating EventNotificationTypes.
+type EventNotificationType struct {
+  unversioned.TypeMeta
+  kapi.ObjectMeta
+  // Description is the human readable description of the notification type.
+  Description string
+}
+
+// UserEventNotification specifies how a user should be notified of a given event.
+type UserEventNotification struct {
+  unversioned.TypeMeta
+  // Name is a User.Name.
+  kapi.ObjectMeta
+  // Notifications map an event type to a list of EventNotificationType references. A user
+  // can receive multiple notifications for an event.
+  Notifications map[string][]ObjectReference
+}
+
+// TODO: define a small mapping API which translates event types to human readable descriptions. The Event.Message field is insufficient because it's object instance-specific.
+```
+
+### Console
+
+With this API, the the Origin console could provide a UI for user notification management by presenting a list of possible `Event` types, and for each type, a selection of `EventNotificationTypes`. If the cluster has no `EventNotificationTypes`, the management feature can be disabled entirely in the console or presented in a way that indicates the cluster doesn't support any notifications.
+
+### CLI
+
+TODO: Write me.
+
+## Use case: Separate email and SMS notification providers
+
+Two notification providers are implemented as applications in the Origin cluster: one for email notification and one for SMS notification, both via third-party services. Each provider lives in its own namespace within the cluster.
+
+In namespace *N*, users `alice` and `bob` want email for `Pod:PodFailed` events, while user `jane` wants both email and text notifications for `DeploymentConfig:Failed`.
+
+The following resources express this cluster configuration:
+
+```yaml
+kind: EventNotificationType
+namespace: email-provider
+name: email
+description: Email notification.
+
+kind: EventNotificationType
+namespace: sms-provider
+name: sms
+description: SMS notification.
+
+kind: UserEventNotification
+name: alice
+notifications:
+  "Pod:PodFailed":
+  - kind: ObjectReference
+    namespace: email-provider
+    name: email
+
+kind: UserEventNotification
```

Comment:
Interesting idea. Would that be abstracted behind the REST API? Applying a defaults from a referenced "default" UserEventNotification when creating new UserEventNotifications?

Otherwise API consumers would have to handle default lookup/merging.